File: workshopIS/JS/Controllers/LoanController.js
var LoanController = function ($scope, DataService) {
    $scope.models = {
        partners: [],
        editedCustomer: null,
        currentLoan: null,
        selectedPartner: null
    };

    // Initialization
    $scope.onInit = function () {
        // set defaults
        $scope.models.editedCustomer = new Customer();
        $scope.models.currentLoan = new Loan();
        $scope.models.currentLoan.interest = .1;
        $scope.models.selectedPartner = null;
        // if data from server has already been fetched
        if (DataService.partners) {
            $scope.models.partners = DataService.partners;
            window.console.log("Loans controller initialized.");
        }
        else
            // fetch data
            DataService.getPartners().then(() => {
            $scope.models.partners = DataService.partners;
            window.console.log("Loans controller initialized.");
        });
    }

    // Link partner to customer, when partner is selected from dropdown
    $scope.onPartnerSelected = function (partner) {
        window.console.log(partner);
        $scope.models.selectedPartner = $scope.models.partners.find((p) => {
            return p.name === partner;
        });
        $scope.models.editedCustomer.partner = $scope.models.selectedPartner;
    }

    // Send form to server and create loan on success
    $scope.onCreateLoan = function (valid) {
        window.console.log($scope.models);
        window.console.log(valid);
        if (valid) {
            window.console.log("Data valid. Creating new loan...");
            DataService.addLoan({
                amount: $scope.models.currentLoan.amount,
                duration: $scope.models.currentLoan.duration,
                interest: $scope.models.currentLoan.interest,
                partnerId: $scope.models.selectedPartner.id,
                phone: $scope.models.editedCustomer.phone,
                firstName: $scope.models.editedCustomer.firstName,
                surname: $scope.models.editedCustomer.surname,
                email: $scope.models.editedCustomer.email,
                note: $scope.models.editedCustomer.note
            });
        }
        else {
            window.console.log("Invalid data! Could not create loan.");
        }
    }

    // Reset form
    $scope.clearForm = function (form) {
        // reset values
        $scope.models.editedCustomer = new Customer();
        $scope.models.currentLoan = new Loan();
        $scope.models.currentLoan.interest = .1;
        $scope.models.selectedPartner = null;
        window.console.log("Clearing form...");
        // reset form
        var controlNames = Object.keys(form).filter(key => key.indexOf('$') !== 0);
        for (var name of controlNames) {
            if (form[name].control)
                form[name].control.$modelValue = null;
        }
        form.$setPristine();
        form.$setUntouched();
    }
}

LoanController.$inject = ["$scope", "DataService"]; 

File: workshopIS/Controllers/LoanController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web.Http;
using System.Web.Mvc;
using System.Web.UI.WebControls;
using workshopIS.Models;

namespace workshopIS.Controllers
{
    // Controller
    public class LoanController : ApiController
    {
        // POST: api/calculator
        /// <summary>
        /// Create new customer from post
        /// </summary>
        /// <param name="loanData">Requested data</param>
        /// <response code="200">Ok</response>
        /// <response code="400">Bad request</response>
        /// <response code="500">Internal Server Error</response>
        /// <returns>Status</returns>
        public IHttpActionResult Post([FromBody]LoanData loanData)
        {
            CCustomer customer;
            CPartner partner;
            CLoan loan;

            if (loanData.partnerId == null)
                return BadRequest("Partner's ID was empty!");

            // Check if partner exists
            if ((partner = (CPartner)Data.Partners.Find(p => p.Id == loanData.partnerId))
                    == null)
                return BadRequest("Could not find selected partner!");

            // Check if Customer already exists
            if ((customer = (CCustomer)partner.Customers
            .Find(c => c.Phone == loanData.phone
                    && c.FirstName == loanData.firstName
                    && c.Surname == loanData.surname)) == null)
            {
                // if Customer doesn't exist
                // Create new Customer... 
                customer = new CCustomer(partner, loanData.phone, 
                    loanData.firstName, loanData.surname, loanData.email);
                // ... and customer to partner
                partner.AddCustomer(customer);
            }
            // ...and new Loan
            loan = new CLoan(customer, loanData.amount, loanData.duration, loanData.interest, loanData.note);
            // add to customer
            customer.AddLoan(loan);

            // return status
            return Ok("Added new loan to customer " +
                (customer.FirstName != null ? customer.FirstName + " " : "") +
                (customer.Surname != null ? customer.Surname + " " : "") +
                customer.Phone + ".\n" +
                "\tLoan:\n" +
                "\t\tAmount:" + loan.Amount + ". Duration: " + loan.Duration + ".\n" +
                "\t\tMonthly charge: " + loan.MonthlyCharge + ".\n" +
                "\t\tAnnual percentage rate:" + loan.APR + ".");
        }

        // GET: api/calculator
        /// <summary>
        /// Display all partners, their customers and their loan data
        /// </summary>
        /// <returns>List of list of customers.</returns>
        public List<IPartner> Get()
        {
            Data.ReadDataFromDatabase();
            return Data.Partners.ToList();
        }
    }

    // Post message class
    public class LoanData
    {
        public int? partnerId;
        public string phone;
        public decimal amount;
        public decimal interest = 0.1M;
        public int duration;
        public string firstName = null;
        public string surname = null;
        public string email = null;
        public string note = null;
    }
}
